samza-core/src/main/java/org/apache/samza/job/metadata/JobCoordinatorMetadataManager.java

```diff
@@ -134,29 +135,34 @@ public JobCoordinatorMetadata generateJobCoordinatorMetadata(JobModel jobModel,
    *
    * @return true if metadata changed, false otherwise
    */
-  public boolean checkForMetadataChanges(JobCoordinatorMetadata newMetadata, JobCoordinatorMetadata previousMetadata) {
-    boolean changed = true;
-
-    if (previousMetadata == null) {
-      metrics.setNewDeployment(1);
-    } else if (!previousMetadata.getEpochId().equals(newMetadata.getEpochId())) {
+  public Set<JobMetadataChange> checkForMetadataChanges(JobCoordinatorMetadata newMetadata,
+      JobCoordinatorMetadata previousMetadata) {
+    Set<JobMetadataChange> changes = new HashSet<>();
+    boolean newDeployment = false;
+    if (previousMetadata == null || !previousMetadata.getEpochId().equals(newMetadata.getEpochId())) {
       metrics.setNewDeployment(1);
-    } else if (!previousMetadata.getJobModelId().equals(newMetadata.getJobModelId())) {
-      metrics.setJobModelChangedAcrossApplicationAttempt(1);
-    } else if (!previousMetadata.getConfigId().equals(newMetadata.getConfigId())) {
-      metrics.setConfigChangedAcrossApplicationAttempt(1);
-    } else {
-      changed = false;
-      metrics.incrementApplicationAttemptCount();
+      newDeployment = true;
+      changes.add(JobMetadataChange.NEW_DEPLOYMENT);
     }
-
-    if (changed) {
-      LOG.info("Job coordinator metadata changed from: {} to: {}", previousMetadata, newMetadata);
-    } else {
+    if (previousMetadata == null || !previousMetadata.getJobModelId().equals(newMetadata.getJobModelId())) {
+      if (!newDeployment) {
+        metrics.setJobModelChangedAcrossApplicationAttempt(1);
+      }
+      changes.add(JobMetadataChange.JOB_MODEL);
+    }
+    if (previousMetadata == null || !previousMetadata.getConfigId().equals(newMetadata.getConfigId())) {
+      if (!newDeployment) {
+        metrics.setConfigChangedAcrossApplicationAttempt(1);
+      }
+      changes.add(JobMetadataChange.CONFIG);
+    }
+    if (changes.isEmpty()) {
       LOG.info("Job coordinator metadata {} unchanged.", newMetadata);
+      metrics.incrementApplicationAttemptCount();
+    } else {
+      LOG.info("Job coordinator metadata changed from: {} to: {}", previousMetadata, newMetadata);
```

Comment:
Thoughts on early termination on `previousMetadata` being `null`? I am not fan of multiple return statements in general but it seems to help w/ readability in this case and eliminates to check for new deployment in every branch.

Reply:
IIRC, I did have that early termination check initially, but I liked how the current way has only a single explicit condition for each different type of change. I was considering the case where someone adds a new `JobMetadataChange`: they will be forced to add a new condition, which will help ensure it gets handled properly. If we use early termination, then a new enum might get automatically lumped in, and that might be undesirable.
I don't think early termination eliminates the check for new deployment, due to the semantics of the metrics. `previousMetadata == null` doesn't cover the case of epoch id changing, but an epoch id change doesn't necessarily always mean a job model change.

Reply:
>doesn't cover the case of epoch id changing, but an epoch id change doesn't necessarily always mean a job model change

Why is this distinction important? How is this being used?

Reply:
This particular case of "epoch id changes but the job model does not change" is not currently needed, but the case of "epoch id does not change and job model changes" is needed. Since this method is being changed to have more granularity, then it seems like also being able to detect "epoch id changes but the job model does not change" is more accurate, in case that distinction is used in the future.

Reply:
hmm.. in that case, i am thinking we should simplify this method and only capture what changes and emit the metrics associated with them and then have the callers/client apply their logic to interpret this logic.
This would mean we will break the invariants defined by the metrics early on although this feature by itself is not stable yet and it should be fine.

Thoughts?

Reply:
Oh ok, if you are fine changing the metrics, then we can do that. Seems a little out-of-scope of the intention of this PR, but I can change the metrics to just emit "job model changed" and "config changed" instead of "job model changed across attempts" and "config changed across attempts". Is that what you were thinking?
If we do this, I still prefer to not short-circuit on `previousMetadata == null`, because then we would need to have multiple places where each metric and each `JobMetadataChange` gets added to the results list.

Reply:
I think we can leave it as is and let me think a bit more about this and follow it up with another PR.